tiledb/sm/filesystem/s3.cc

```diff
@@ -461,9 +461,18 @@ Status S3::ls(
     }
 
     is_done = !list_objects_outcome.GetResult().GetIsTruncated();
-    if (!is_done)
-      list_objects_request.SetMarker(
-          list_objects_outcome.GetResult().GetNextMarker());
+    if (!is_done) {
+      // The documentation states that "GetNextMarker" will be non-empty only
+      // when the delimiter in the request is non-empty. When the delimiter is
+      // non-empty, we must used the last returned key as the next marker.
+      std::string next_marker =
+          !delimiter.empty() ?
+              list_objects_outcome.GetResult().GetNextMarker() :
+              list_objects_outcome.GetResult().GetContents().back().GetKey();
```

Comment:
What happens with `list_objects_outcome.GetResult().GetContents().back()` if the response was empty?

Reply:
I believe this to be part of the contract: If the returned object list was truncated, it implies that it was too big to fit into a single response and therefore should have at least one object. I don't see any harm in assert()ing this assumption, or returning an non-ok Status. Which do you prefer?

Reply:
If it's part of the SDK contract, an assert is fine with me.